### PR TITLE
FPU specifications

### PR DESCRIPTION
**Pull request by JonathanSalwan**

### CODE REVIEW

**Comment by JonathanSalwan** [comment]
#875

File: src/libtriton/utils/coreUtils.cpp [hunk 1]
```diff
@@ -17,6 +17,14 @@
 namespace triton {
   namespace utils {
 
+    void fromUintToBuffer(triton::uint80 value, triton::uint8* buffer) {
```
[the remark this answers is not in the document]
Comment (fvrmatteo) [reply src/libtriton/utils/coreUtils.cpp:20]: Added.

File: src/libtriton/engines/symbolic/symbolicEngine.cpp [hunk 2]
```diff
@@ -196,6 +197,7 @@ namespace triton {
           if (index < triton::size::word)    this->alignedMemoryReference.erase(std::make_pair(address-index, triton::size::word));
           if (index < triton::size::dword)   this->alignedMemoryReference.erase(std::make_pair(address-index, triton::size::dword));
           if (index < triton::size::qword)   this->alignedMemoryReference.erase(std::make_pair(address-index, triton::size::qword));
+          if (index < triton::size::qword)   this->alignedMemoryReference.erase(std::make_pair(address-index, triton::size::fword));
```
[the remark this answers is not in the document]
Comment (fvrmatteo) [reply src/libtriton/engines/symbolic/symbolicEngine.cpp]: Fixed.

File: src/libtriton/includes/triton/x86.spec [hunk 3]
```diff
@@ -200,14 +219,18 @@ REG_SPEC(CR14, cr14, triton::bitsize::qword-1, 0, CR14, triton::bitsize::dword-1
 REG_SPEC(CR15, cr15, triton::bitsize::qword-1, 0, CR15, triton::bitsize::dword-1, 0, CR15, true)  // cr15
 
 /* Debug */
+
 REG_SPEC(DR0,  dr0,  triton::bitsize::qword-1, 0, DR0,  triton::bitsize::dword-1, 0, DR0,  true)  // dr0
 REG_SPEC(DR1,  dr1,  triton::bitsize::qword-1, 0, DR1,  triton::bitsize::dword-1, 0, DR1,  true)  // dr1
 REG_SPEC(DR2,  dr2,  triton::bitsize::qword-1, 0, DR2,  triton::bitsize::dword-1, 0, DR2,  true)  // dr2
 REG_SPEC(DR3,  dr3,  triton::bitsize::qword-1, 0, DR3,  triton::bitsize::dword-1, 0, DR3,  true)  // dr3
+REG_SPEC(DR4,  dr4,  triton::bitsize::qword-1, 0, DR4,  triton::bitsize::dword-1, 0, DR4,  true)  // dr4
+REG_SPEC(DR5,  dr5,  triton::bitsize::qword-1, 0, DR5,  triton::bitsize::dword-1, 0, DR5,  true)  // dr5
```
[the remark this answers is not in the document]
Comment (fvrmatteo) [reply src/libtriton/includes/triton/x86.spec]: Removed, I didn't know they were obsolete synonyms :)

File: src/testers/x86/unicorn_test_x86.py [hunk 4]
```diff
@@ -744,6 +744,11 @@
     (None, b"\x48\xc7\xc0\xff\xff\xff\xff",               "mov rax, 0xffffffffffffffff"),
     (None, b"\x48\xc7\xc3\x00\x00\x00\x00",               "mov rbx, 0x0"),
     (None, b"\xf3\x0f\xbc\xd8",                           "tzcnt ebx, eax"),
+
+    (None, b"\x0f\xae\x04\x24",                           "fxsave [rsp]"),
+    (None, b"\x0f\xae\x0c\x24",                           "fxrstor [rsp]"),
+    (None, b"\x48\x0f\xae\x04\x24",                       "fxsave64 [rsp]"),
+    (None, b"\x48\x0f\xae\x0c\x24",                       "fxrstor64 [rsp]"),
```
[the remark this answers is not in the document]
Comment (fvrmatteo) [reply src/testers/x86/unicorn_test_x86.py]: I tried adding more tests and I think the inconsistencies between Unicorn and a real execution are too heavy to get anything useful out of these comparisons (unless we find a way to get Unicorn to behave correctly). The MMX registers on Unicorn are always 0 and the fxsave is not behaving like a real CPU is behaving. The reserved bytes are also vendor dependent and Unicorn is clearing some of them with 0 and others not.

[the remark this answers is not in the document]
Comment (fvrmatteo) [reply src/testers/x86/unicorn_test_x86.py]: I also realised that all the instructions modifying the MMX registers will need to call the new **updateFTW** function to update the state of the x87 FPU Tag Word and make it consistent with a real execution. I'll take care of adding the call on all the semantics stubs that update an MMX register.

**Comment by fvrmatteo** [comment]
- I still have to remove the ugly macros, I'm keeping them there until I see that everything is working and then I'll preprocess them. I can confirm that the **std::memcpy** doesn't work on the Boost 80/128/256/512 types, so I'm using the **fromUintToBuffer** and **fromBufferToUint** functions now and I can also confirm that the read/write to the registers using those types it's correct (and we don't incur into wrong type punning issues).
- **SymbolicEngine::insertSubRegisterInParent** was already handling the fword case, but please if you think something is missing there let me know.

**Comment by JonathanSalwan** [comment]
> SymbolicEngine::insertSubRegisterInParent was already handling the fword case, but please if you think something is missing there let me know.

I think we just have to put the fword case like:

```patch
   case triton::size::dword:  
   case triton::size::qword:  
+  case triton::size::fword:  
   case triton::size::dqword: 
   case triton::size::qqword: 
   case triton::size::dqqword:

```

**Comment by fvrmatteo** [comment]
The latest commit takes care of:
1. removing the ugly macros;
2. handling the fword size properly;
3. updating the x87 FPU Tag Word each time that an MMX register is modified;
4. disabling the fxsave/fxsave64/fxrstor/fxrstor64 Unicorn tests because the MMX registers are unsupported and the reserved memory doesn't match with a real execution.

In the x87 FPU Tag Word calculation the tainting is currently disabled because I don't know if it makes sense to have it.

**Comment by fvrmatteo** [comment]
Apparently the tests are failing but I'm not sure why, I think it is related to the definition of the new registers on the **x86.spec** file but I'll wait some feedback to know how to proceed.

**Comment by fvrmatteo** [comment]
@JonathanSalwan The build seems to be passing the tests now, although Travis is also failing while compiling Unicorn because it doesn't find QEMU, but that's an unrelated issue it seems. With the introduction of the STX registers (extension to 80-bits of the MMX registers) and the fixing of some bit flags everything seems to be passing the checks.

File: src/libtriton/CMakeLists.txt [hunk 5]
```diff
@@ -49,6 +49,9 @@ if(CMAKE_COMPILER_IS_GNUCXX OR ${CMAKE_CXX_COMPILER_ID} MATCHES "Clang")
     endif()
 endif()
 
+# Enable preprocessor output
+# set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --save-temps")
+
```
Comment (JonathanSalwan) [inline src/libtriton/CMakeLists.txt:54]: Can you remove these lines?

**Comment by SweetVishnya** [comment]
Any update on this pull request? Supporting FPU instructions would be a great feature.

**Comment by fvrmatteo** [comment]
@SweetVishnya @JonathanSalwan I would gladly come back to this if we can somehow organise to come up with proper testing, verify the code and decide how to proceed. I can try to update the PR to support the latest version of Triton. Testing-wise we still have the issue with Unicorn that doesn't properly support some registers and differs on some vendor-specific effects of the FXSAVE/FXRSTOR instructions.

**Comment by JonathanSalwan** [comment]
What about using a DBI to test the semantics (at least for x86) ?

**Comment by fvrmatteo** [comment]
@JonathanSalwan Could you please point me to an example of what you mean with using a DBI to test the semantics? I would say that as long as we can dump the registers and memory values to be compared we are fine.

**Comment by mrexodia** [comment]
Any movement on this? I resolved the conflicts locally and they all seem trivial. Perhaps it would be good to merge this and then open an issue for the tests after that?

**Comment by SweetVishnya** [comment]
Untested instructions may break a path predicate. So, previously correct solutions for already existing instructions may become incorrect or even unsatisfiable?

@JonathanSalwan, could u create a branch with this features merged with dev-v0.9?

On our side, we can run benchmarking on real programs. Thus, we can measure correctness of inverted branches.

**Comment by SweetVishnya** [comment]
@fvrmatteo, there is also the following method to test it. You can write a C program that performs some fp computations and does printing inside branches. Thus, you can ask Triton to solve a path predicate, run the program on generated input, and verify that program has an expected output.

**Comment by mrexodia** [comment]
> Untested instructions may break a path predicate. So, previously correct solutions for already existing instructions may become incorrect or even unsatisfiable.
> 
> @JonathanSalwan, could u create a branch with this features merged with dev-v0.9?
> 
> On our side, we can run benchmarking on real programs. Thus, we can measure correctness of inverted branches.

I really don't see how additional semantics could break existing things. Previously there was no semantic for the instructions and none of the new ones have a condition that influences the instruction pointer. Perhaps I am missing something though...

**Comment by SweetVishnya** [comment]
They can lead to unsatisfiable path constraints. Those constraints conjuncted with previous branches will make some paths infeasible.

**Comment by mrexodia** [comment]
Could you give a concrete example of what that would look like?

**Comment by SweetVishnya** [comment]
int a, b, c;
if (a + b + c != 1.0) { /* 1 */ }
if (c == 3) { /* 2 */ }

If integer to fp conversation contains error, 2 may become infeasible.

**Comment by mrexodia** [comment]
Okay, but this wouldn’t have worked in the first place because of missing semantics right?

**Comment by SweetVishnya** [comment]
That was working before. Two branches are independent.

**Comment by SweetVishnya** [comment]
The main thing about symbolic reasoning. When you add more constraints, you should test that they do not break data and control flow.

**Comment by fvrmatteo** [comment]
@SweetVishnya Sorry for the late answer, I'm also confused here!

The original discussion was about testing the semantic of the newly added instructions (fxsave/fxsave64/fxrstor/fxrstor64), specifically the impossibility to test them using Unicorn because the concrete memory state was wrong.

The semantic of the other FPU instructions didn't change and the new aforementioned instructions are not adding path constraints, so how is the latest example relevant and why the testing shifted to a path constraint problem from the original issue?

The tests for the existing FPU instructions should still be working after this patch and a possible strategy to test the semantic of the new instructions could be using something like microx (from Trail Of Bits) to execute them on the machine and compare the memory values (ignoring the vendor specific bits).

**Comment by SweetVishnya** [comment]
Sorry, seems like I didn't get what this merge is about. I suppose that simple moves won't break anything. I am just little worried about commented assertions checking sizes.

After this merge Triton won't still be able to perform FPU stack arithmetic computations. Am I correct?

**Comment by SweetVishnya** [comment]
I looked that up. Triton does not implement `fadd`. Again, forgive me for confusion, please.

**Comment by fvrmatteo** [comment]
@SweetVishnya Don't worry, glad we managed to clarify the confusion! :) I'm now working with @mrexodia to see if we can come up with proper testing for the newly added instructions (and we'll take a look at the commented out checks too).

About the missing semantic you are mentioning (`fadd`), they'll need to be added with a separate PR I'd say. Other projects are implementing many more FPU semantics and with the support to the new registers added by this PR it should be possible to port them.

**Comment by JonathanSalwan** [comment]
Thanks for the rebase to `dev-fpu`. I will review the MR sunday (tomorrow) and will merge it to `dev-fpu`. Once it's up-to-date with `dev-v0.9` we can start making unittests for them. For example something like doing an `XSAVE`, then trashing fp registers, then doing a `XRSTORE` and then printing all registers. If we have the same output from a real execution and the Triton emulation we can say that our implementation is correct. What do you think?

**Comment by mrexodia** [comment]
This branch is based off master, so I would suggest merging into dev-v0.9
(from what I can tell dev-fpu is a dead branch ages behind)

**Comment by JonathanSalwan** [comment]
I've just rebased `dev-fpu` to `dev-v0.9`. Should be easier for me to review the MR :)

**Comment by mrexodia** [comment]
That works too I guess. I don’t really see the point in having yet another branch away from master, but it’s your project ^^

**Comment by JonathanSalwan** [comment]
Once unittests are made I can merge it to dev-v0.9 :). It's just to keep the branch v0.9 as stable as I can

**Comment by JonathanSalwan** [comment]
Thanks for this MR, it's merged. If someone has a good idea for the unittest part go ahead :).

**Comment by JonathanSalwan** [comment]
Once we are 100% sure of our implementation i will merge it to dev-v0.9

**Pull request merged**